File: 96_Word/ruby/word.rb
#!/usr/bin/env ruby
# WORD
#
# Converted from BASIC to Ruby


WORDS = ["DINKY", "SMOKE", "WATER", "GRASS", "TRAIN", "MIGHT",
         "FIRST","CANDY", "CHAMP", "WOULD", "CLUMP", "DOPEY"]

def game_loop
  target_word = WORDS.sample.downcase
  guess_count = 0
  guess_progress = ["-"] * 5

  puts "You are starting a new game..."
  while true
    guess_word = ""
    while guess_word == ""
      puts "Guess a five letter word. "
      guess_word = gets.chomp
      if guess_word == "?"
        break
      elsif !guess_word.match(/^[[:alpha:]]+$/) || guess_word.length != 5
        guess_word = ""
        puts "You must guess a five letter word. Start again."
      end
    end
    guess_count += 1
    if guess_word == "?"
      puts "The secret word is #{target_word}"
      break
    else
      common_letters = ""
      matches = 0
      5.times do |i|
        5.times do |j|
          if guess_word[i] == target_word[j]
            matches += 1
            common_letters = common_letters + guess_word[i]
            guess_progress[j] = guess_word[i] if i == j
          end
        end
      end
      puts "There were #{matches} matches and the common letters were... #{common_letters}"
      puts "From the exact letter matches, you know............ #{guess_progress.join}"
      if guess_progress.join == guess_word
        puts "You have guessed the word. It took #{guess_count} guesses!"
        break
      elsif matches < 2
        puts "If you give up, type '?' for you next guess."
      end
    end
  end
end

puts " " * 33 + "WORD"
puts " " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n"
puts "I am thinking of a word -- you guess it. I will give you"
puts "clues to help you get it. Good luck!!\n"

keep_playing = true
while keep_playing
  game_loop
  puts "\n Want to play again? "
  keep_playing = gets.chomp.downcase.index("y") == 0
end

File: 96_Word/python/word.py
#!/usr/bin/env python3
# WORD
#
# Converted from BASIC to Python by Trevor Hobson

import random

words = ["DINKY", "SMOKE", "WATER", "GRASS", "TRAIN", "MIGHT", "FIRST",
         "CANDY", "CHAMP", "WOULD", "CLUMP", "DOPEY"]


def play_game():
    """Play one round of the game"""

    random.shuffle(words)
    target_word = words[0]
    guess_count = 0
    guess_progress = ["-"] * 5

    print("You are starting a new game...")
    while True:
        guess_word = ""
        while guess_word == "":
            guess_word = input("\nGuess a five letter word. ").upper()
            if guess_word == "?":
                break
            elif not guess_word.isalpha() or len(guess_word) != 5:
                guess_word = ""
                print("You must guess a five letter word. Start again.")
        guess_count += 1
        if guess_word == "?":
            print("The secret word is", target_word)
            break
        else:
            common_letters = ""
            matches = 0
            for i in range(5):
                for j in range(5):
                    if guess_word[i] == target_word[j]:
                        matches += 1
                        common_letters = common_letters + guess_word[i]
                        if i == j:
                            guess_progress[j] = guess_word[i]
            print("There were", matches,
                  "matches and the common letters were... " + common_letters)
            print(
                "From the exact letter matches, you know............ " + "".join(guess_progress))
            if "".join(guess_progress) == guess_word:
                print("\nYou have guessed the word. It took",
                      guess_count, "guesses!")
                break
            elif matches == 0:
                print("\nIf you give up, type '?' for you next guess.")


def main():
    print(" " * 33 + "WORD")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n")

    print("I am thinking of a word -- you guess it. I will give you")
    print("clues to help you get it. Good luck!!\n")

    keep_playing = True
    while keep_playing:
        play_game()
        keep_playing = input(
            "\nWant to play again? ").lower().startswith("y")


if __name__ == "__main__":
    main()
